src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,176 @@
+import java.util.HashMap;
+
+import org.usfirst.frc.team2473.framework.Devices;
+import org.usfirst.frc.team2473.robot.RobotMap;
+
+import com.ctre.phoenix.ErrorCode;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+
+public class DriveSubsystem extends Subsystem {
+	
+	private static DriveSubsystem instance;
+	
+	static {
+		instance = new DriveSubsystem();
+	}
+	
+	public static DriveSubsystem getInstance() {
+		return instance;
+	}
+	
+	WPI_TalonSRX backLeft;
+	WPI_TalonSRX backRight;
+	WPI_TalonSRX frontLeft;
+	WPI_TalonSRX frontRight;
+	
+	DifferentialDrive teleopDrive;
+	
+	private HashMap<Double, Double> leftTable = new HashMap<>();
+	private HashMap<Double, Double> rightTable = new HashMap<>();
+	private double minTestedPower, maxTestedPower;
+	
+	private DriveSubsystem() {
+		SpeedControllerGroup right = new SpeedControllerGroup(Devices.getInstance().getTalon(RobotMap.TALON_BR), Devices.getInstance().getTalon(RobotMap.TALON_FR));
+		SpeedControllerGroup left = new SpeedControllerGroup(Devices.getInstance().getTalon(RobotMap.TALON_BL), Devices.getInstance().getTalon(RobotMap.TALON_FL));
+		teleopDrive = new DifferentialDrive(right, left);
+		
+		backLeft = Devices.getInstance().getTalon(RobotMap.TALON_BL);
+		backRight = Devices.getInstance().getTalon(RobotMap.TALON_BR);
+		frontLeft = Devices.getInstance().getTalon(RobotMap.TALON_FL);
+		frontRight = Devices.getInstance().getTalon(RobotMap.TALON_FR);
+		initLookupTable();
+	}
+	
+	private void initLookupTable() {
+		leftTable.put(0.2, 1.035842);
+		leftTable.put(0.3, 1.033883);
+		leftTable.put(0.4, 1.044270);
+		leftTable.put(0.5, 1.055146);
+		leftTable.put(0.6, 1.052553);
+		leftTable.put(0.7, 1.061291);
+		leftTable.put(0.8, 1.088878);
+		
+		rightTable.put(0.2, 1.012183);
+		rightTable.put(0.3, 1.017407);
+		rightTable.put(0.4, 1.021134);
+		rightTable.put(0.5, 1.027404); // was 1.025404
+		rightTable.put(0.6, 1.020292);
+		rightTable.put(0.7, 1.017890);
+		rightTable.put(0.8, 1.033552);
+		
+		minTestedPower = Collections.min(leftTable.keySet());
+		maxTestedPower = Collections.max(leftTable.keySet());
+	}
+	
+	public double convertPower(double power, boolean isLeft) {
+		HashMap<Double, Double> tempTable = isLeft ? leftTable : rightTable;
+		if (power < minTestedPower) {
+			return power * 1/tempTable.get(minTestedPower);
+		} else if (power > maxTestedPower) {
+			return power * 1/tempTable.get(maxTestedPower);
+		} else {
+			// linearize between the two values around it
+			ArrayList<Double> powers = new ArrayList<>(tempTable.keySet());
+			Collections.sort(powers);
+			
+			if (powers.contains(power)) { // the input power is one of the powers in the lookup table
+				double newPower = power * 1/tempTable.get(power);
+				return newPower;
+			}
+			
+			int i = 0;
+			while(powers.get(i) < power) i++;
```

Comment:
I challenge you to rewrite this as a one line for loop. Also what is up with that `i--`?

Reply:
Also this wasn't written as a for loop because we need to access ```i``` afterwards.

Reply:
You can declare int i; ahead of the loop so it is within the entire function's scope.